### PR TITLE
JetBrains: Make auth more convenient

### PR DESCRIPTION
**Pull request by vdavid**
Closes https://github.com/sourcegraph/sourcegraph/issues/36783

The work items here were these:
1. Add "User authentication" and "Navigation settings" groups
2. Add help texts
3. Set up URL syntax validation
4. Set up automatic access token validation after an access token change or URL change

--- 

Details on each:

**1. Grouping:** It was a pain. It was rather difficult to figure it out without the UI Inspector, which I couldn't make work at first. But then today morning I managed to pop it up: I still don't have a menu item for it, but `⌃⌥click` brings it up in my sandboxed instance! So that made it somewhat quicker, and I've added the two groups and the indentation to make the two groups of settings look somewhat like JetBrains' UI guidelines.

**2. Help texts:** I've added many of them. They don't closely follow the [figma](https://www.figma.com/file/nmNPZGYCoGIUN0YMylmPTT/JetBrains-plugin-MLP?node-id=246%3A11) but we discussed with @jjinnii that even the figma versions are not final and let's not get blocked by the final copy, as it's something that's relatively easy to change later.

**3. URL validation:** Turns out that JetBrains has pretty good [guidelines](https://jetbrains.design/intellij/principles/validation_errors/) for this, that even includes some code! The code is somewhat obsolete and it didn't really work for me out of the box, but I found a way and now it's very stable. I also extracted the most verbose parts to private methods, so now we can add validations with lambda functions in a much more concise way. The validation itself is not my custom one but one that's part of `JsonSchemaConfigurable`, as this seemed to be the most standard one. TBH, it's cr*p, it basically notices if a URL doesn't start with "http", but accepts pretty much everything else. I think this is a good baseline of defense, though.

**4. Token validation:** I only implemented a simplistic version. I think this is acceptable, but I'm also interested in other opinions. Currently, the validation only considers the field valid if it's either exactly 40 characters long, or empty. This should cover most honest mistakes. The next step here would be to implement remote validation, but:
 - Currently, we only have the access token validation code on the JS side, and adding it to the Java side would result in some duplication of functionality, and would take time.
 - I already took so long with this because of the lack of documentation on Swing and IntelliJ components that I'd prefer to cut this effort short and go with this more simplistic validation. I honestly think it helps users with the most common mistakes like not copying the first/last character of the token.

## Test plan

- Here is the UI demo: [3-min Loom](https://www.loom.com/share/a52b6566f88b4a1a8e0f6e9ae53c493a)
- Check the code

## App preview:

- [Web](https://sg-web-dv-jetbrains-make-auth-more.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-appvhjimib.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

File: client/jetbrains/src/main/java/com/sourcegraph/config/SettingsComponent.java [hunk 1]
```diff
@@ -123,4 +105,116 @@ public boolean isUrlNotificationDismissed() {
...
+
+        JBLabel accessTokenLabel = new JBLabel("Access token:");
+        accessTokenTextField = new JBTextField();
+        accessTokenTextField.getEmptyText().setText("Paste your access token here");
+        addValidation(accessTokenTextField, () ->
+            (accessTokenTextField.getText().length() > 0 && accessTokenTextField.getText().length() != 40)
+                ? new ValidationInfo("Invalid access token", accessTokenTextField)
+                : null);
+
+        JBLabel userDocsLinkComment = new JBLabel("<html><body>You might need an access token to sign in. See our <a href=\"https://docs.sourcegraph.com/cli/how-tos/creating_an_access_token\">user docs</a> for a video guide,</body></html>", UIUtil.ComponentStyle.SMALL, UIUtil.FontColor.BRIGHTER);
```
Comment (philipp-spiess) [inline client/jetbrains/src/main/java/com/sourcegraph/config/SettingsComponent.java:131]: Lol of course this is a full HTML document 🙈  I bet this was very intuitive!

**Review by philipp-spiess — APPROVED** [review]
Good stuff! Reading thought he code changes, I felt the pain that you went through though 😢 

Oprah Crazy

**Review by jjinnii — APPROVED** [review]
Looks amazing, thank you so much @vdavid for your perseverance on this issue! 

Just two small nits on copy (non-blocking to merge): 

- Is there a reason we say "you _might_ need an access token"? We try to avoid any directions that aren't clear, e.g. in what situation will they need or not need an access token? 
- Instead of "never show the 'no sourcegraph url'...", can we replace with "Do not show...". That's more in line with copy standards.

File: client/jetbrains/src/main/java/com/sourcegraph/config/SettingsComponent.java (hunk 1 above)
Comment (vdavid) [reply client/jetbrains/src/main/java/com/sourcegraph/config/SettingsComponent.java:131]: Yeah, if you just put HTML tags in there, it doesn't work. Caused me some time to find this out indeed 🤦‍♂️

**Comment by vdavid** [comment]
@jjinnii 

> Is there a reason we say "you might need an access token"? We try to avoid any directions that aren't clear, e.g. in what situation will they need or not need an access token?

Ryan discovered that some of our enterprise clients have proxies in their system that add the access token to Sourcegraph requests on the fly, automatically.
Imagine that your computer sends the request _without_ a token, but the request goes through a proxy, and when the proxy sees that it's a legit request from a company machine, on its way to the company's legit Sourcegraph instance, then it automatically adds the access token to the request header.
In the case of these users, they have no access to a custom token, they don't need to enter an access token, and they won't even have to know about the existence of access tokens basically. It just works for them as soon as they enter the right URL.
This is the case I mentioned in a couple of my previous comments.
Considering this, I have no better idea but to say "might", but we can change it if you have a better idea!

> Instead of "never show the 'no sourcegraph url'...", can we replace with "Do not show...". That's more in line with copy standards.

Changed it, thank you! :)

**Comment by jjinnii** [comment]
> In the case of these users, they have no access to a custom token, they don't need to enter an access token, and they won't even have to know about the existence of access tokens basically.

@vdavid Do you have any idea how many of our customers this is? 🤔  And I'm assuming it would be a decent amount of work to identify these users and then to disable the access token field for them?

**Comment by philipp-spiess** [comment]
> And I'm assuming it would be a decent amount of work to identify these users and then to disable the access token field for them?

I don't think we can do that ahead of time, no :( Here's some documentation of this feature for the curious: https://docs.sourcegraph.com/admin/auth#http-authentication-proxies

What we could do is to make a HTTP request to the URL once it is entered and see if we have access to a user but I'd say that this is better treated as a post release addition since it adds a lot of complexity.

**Comment by vdavid** [comment]
@ryankscott never mind! We discussed and closed this on Slack.

**Pull request merged**